### PR TITLE
Introduce builtin to access percpu kernel data

### LINKED ISSUE #1837: Support accessing percpu kernel data
There's a helper to access percpu data now: https://github.com/torvalds/linux/blob/0aa099a312b6323495a23d758009eb7fc04a7617/include/uapi/linux/bpf.h#L4489

However, it's not so simple to use. IIUC, there needs to be BTF ID to the data in question
in the right register when the helper is called. libbpf does these fixups and they look
somewhat complicated. So before we can support this we'll need to have libbpf do
our loading for us.

### PR DESCRIPTION
**Pull request by viktormalik**
Introduce new builtin `percpu_kaddr` to access kernel percpu variables. It leverages the `bpf_per_cpu_ptr` helper which accepts a pointer to a percpu ksym and a CPU number. The ksym is an extern variable with a BTF entry matching the BTF of the corresponding kernel variable. Since we now use libbpf to do the loading, it is sufficient to emit a global variable declaration with a proper BTF (debug info) and libbpf will take care of the rest.

The helper has two forms:

    percpu_kaddr("symbol_name")
    percpu_kaddr("symbol_name", N)

where N is the CPU number. The former variant retrieves the value for the current CPU.

For instance, this allows to access the `process_counts` percpu variable:
```
$ bpftrace -e 'BEGIN { printf("%d processes running on CPU %d\n", *percpu_kaddr("process_counts"), cpu); exit() }'
Attaching 1 probe...
88 processes running on CPU 1
```

 A tricky part is that `bpf_per_cpu_ptr` may return NULL if the supplied CPU number is higher than the number of the CPUs. The BPF program should perform a NULL-check on the returned value, otherwise it is rejected by the verifier. In practice, this only happens if pointer arithmetics is used (i.e. a struct field is accessed). Since it is quite complex to detect a missing NULL-check in bpftrace, we instead let verifier do it and just display the potential verifier error in a nicer manner:

```
$ bpftrace -e 'BEGIN { printf("%d\n", ((struct rq *)percpu_kaddr("runqueues"))->nr_running); exit() }'
Attaching 1 probe...
stdin:1:24-64: ERROR: helper bpf_per_cpu_ptr: result needs to be null-checked before accessing fields
BEGIN { printf("%d\n", ((struct rq *)percpu_kaddr("runqueues"))->nr_running); exit() }
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Resolves #1837.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by ajor** [comment]
Since I don't generally like adding new builtins, I've got to ask: what are the benefits of this over reading from the kernel PERCPU data manually, a la: https://github.com/hodgesds/scx/blob/912d6e01c1538cb5ec7e6363332caef66f7765ba/scheds/rust/scx_layered/integration/layer_llc.bt#L22-L27

No bpftrace changes requires and with user defined functions, the above approach could be wrapped in a simple stdlib function to make it easier to use.

**Comment by viktormalik** [comment]
> Since I don't generally like adding new builtins, I've got to ask: what are the benefits of this over reading from the kernel PERCPU data manually, a la: https://github.com/sched-ext/scx/blob/912d6e01c1538cb5ec7e6363332caef66f7765ba/scheds/rust/scx_layered/integration/layer_llc.bt#L22-L27
> 
> No bpftrace changes requires and with user defined functions, the above approach could be wrapped in a simple stdlib function to make it easier to use.

Mostly convenience, at the moment. It will take a while until we get user-defined functions and a standard library and writing this manually is much more code.

On top of that, we get some verifier checks for free, e.g. the mentioned check that users cannot do pointer arithmetic on the returned value without a prior null-check. For stdlib bpftrace function, we'd have to define those checks on our own.

That being said, I understand you reluctance to add new helpers but I feel like the standard library is still quite some time away and, in the meantime, we'll need to address some real-world use-cases which require new builtins. I'm wondering if we could establish a plan that once the standard library is in place, we will remove some (potentially many) builtin functions in favor of stdlib functions? That would allow us to add builtins for the time being but we would also be transparent about the fact that they will go away from the language eventually.

**Comment by jordalgo** [comment]
> I'm wondering if we could establish a plan that once the standard library is in place, we will remove some (potentially many) builtin functions in favor of stdlib functions? 

Is it possible for this to be seamless for users? Maybe we could detect what "builtin" functions are being used and automatically include them from the standard library (if there are no conflicts with other user-defined functions)

**Comment by viktormalik** [comment]
> > I'm wondering if we could establish a plan that once the standard library is in place, we will remove some (potentially many) builtin functions in favor of stdlib functions?
> 
> Is it possible for this to be seamless for users? Maybe we could detect what "builtin" functions are being used and automatically include them from the standard library (if there are no conflicts with other user-defined functions)

Yeah, something like that. It very much depends on the include model that we end up with but I'm sure we'll be able to find a solution which is transparent for users.

File: src/ast/irbuilderbpf.cpp [hunk 1]
```diff
@@ -2141,6 +2141,23 @@ CallInst *IRBuilderBPF::CreateGetFuncIp(Value *ctx, const location &loc)
...
+  //    cpu, or NULL, if cpu is invalid.
+  FunctionType *percpuptr_func_type = FunctionType::get(
+      GET_PTR_TY(), { GET_PTR_TY(), getInt64Ty() }, false);
+  return CreateHelperCall(libbpf::BPF_FUNC_per_cpu_ptr,
+                          percpuptr_func_type,
+                          { var, cpu },
+                          "per_cpu_ptr",
+                          &loc);
+}
+
```
Comment (danobi) [inline src/ast/irbuilderbpf.cpp:2160]: Looks like the `bpf_this_cpu_ptr()` variant can never return null according to docs. Given that user needs to do null checks if accessing fields, probably worthwhile to use the `this` variant when possible. So less null checks necessary. WDYT?

File: src/ast/passes/codegen_llvm.cpp [hunk 2]
```diff
@@ -4697,4 +4710,37 @@ CallInst *CodegenLLVM::CreateKernelFuncCall(Kfunc kfunc,
...
+  if (auto *sym = module_->getGlobalVariable(var_name))
+    return sym;
+
+  std::string err;
+  auto type = bpftrace_.btf_->get_var_type(var_name);
+  if (type.IsNoneTy()) {
+    throw FatalUserException("Could not resolve variable \"" + var_name +
+                             "\" from BTF");
+  }
+
```
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:4734]: Can this be done in semantic analyser so there's highlighting?

File: tests/runtime/call
Comment (danobi) [inline tests/runtime/call:1]: test missing symbol also?

File: src/ast/irbuilderbpf.cpp (hunk 1 above)
Comment (viktormalik) [reply src/ast/irbuilderbpf.cpp:2160]: Yeah, absolutely, I missed that one.

File: src/ast/passes/codegen_llvm.cpp (hunk 2 above)
Comment (viktormalik) [reply src/ast/passes/codegen_llvm.cpp:4734]: Right, makes sense.

File: man/adoc/bpftrace.adoc [hunk 3]
```diff
@@ -1846,6 +1850,39 @@ the path will be clamped by `size` otherwise `BPFTRACE_MAX_STRLEN` is used.
...
+* `void *percpu_kaddr(const string name, int cpu)`
+
+*sync*
+
+Get the address of the percpu kernel symbol `name` for CPU `cpu`. When `cpu` is
+omitted, the current CPU is used.
+
+----
+interval:s:1 {
+  $proc_cnt = kaddr("process_counts");
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:1867]:
```suggestion
  $proc_cnt = percpu_kaddr("process_counts");
```

File: man/adoc/bpftrace.adoc [hunk 4]
```diff
@@ -1846,6 +1850,39 @@ the path will be clamped by `size` otherwise `BPFTRACE_MAX_STRLEN` is used.
...
+
+The function may return NULL if `cpu` is higher than the number of available
+CPUs. Therefore, it is necessary to perform a NULL-check on the result when
+accessing fields of the pointed structure, otherwise the BPF program will be
+rejected.
+
+----
+interval:s:1 {
+  $runqueues = (struct rq *)percpu_kaddr("runqueues");
+  if ($runqueues != 0) {         # The check is mandatory here
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:1880]: Should we add a NULL type ? 🤔

File: man/adoc/bpftrace.adoc [hunk 5]
```diff
@@ -1846,6 +1850,39 @@ the path will be clamped by `size` otherwise `BPFTRACE_MAX_STRLEN` is used.
...
+interval:s:1 {
+  $proc_cnt = kaddr("process_counts");
+  printf("% processes are running on CPU %d\n", *$proc_cnt, cpu);
+}
+----
+
+The function may return NULL if `cpu` is higher than the number of available
+CPUs. Therefore, it is necessary to perform a NULL-check on the result when
+accessing fields of the pointed structure, otherwise the BPF program will be
+rejected.
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:1875]: What happens if the kernel symbol is invalid? e.g. `percpu_kaddr("potato")`

File: src/ast/irbuilderbpf.cpp (hunk 1 above)
Comment (jordalgo) [reply src/ast/irbuilderbpf.cpp:2160]: I wonder if we can bake this into a semantic_analyser check. It's a bit confusing to users that they have to null check one variant and not the other.

File: man/adoc/bpftrace.adoc (hunk 5 above)
Comment (viktormalik) [reply man/adoc/bpftrace.adoc:1875]: Then we fail in the semantic analyser, see one of the unit tests.

File: tests/codegen/llvm/call_len_map_sum_elem_count.ll [hunk 6]
```diff
@@ -117,5 +117,5 @@ attributes #1 = { nocallback nofree nosync nounwind willreturn memory(argmem: re
 !61 = !DISubroutineType(types: !62)
 !62 = !{!18, !63}
 !63 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: !64, size: 64)
-!64 = !DICompositeType(tag: DW_TAG_structure_type, name: "struct bpf_map", scope: !2, file: !2, elements: !65)
+!64 = !DICompositeType(tag: DW_TAG_structure_type, name: "bpf_map", scope: !2, file: !2, elements: !65)
```
Comment (jordalgo) [inline tests/codegen/llvm/call_len_map_sum_elem_count.ll:120]: this a safe change?

File: src/ast/dibuilderbpf.cpp [hunk 7]
```diff
@@ -212,8 +212,14 @@ DIType *DIBuilderBPF::CreateByteArrayType(uint64_t num_bytes)
 DIType *DIBuilderBPF::GetType(const SizedType &stype, bool emit_codegen_types)
 {
   if (!emit_codegen_types && stype.IsRecordTy()) {
+    std::string name = stype.GetName();
+    if (name.find("struct ") == 0)
+      name = name.substr(std::string("struct ").length());
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/ast/dibuilderbpf.cpp:217]: Yeah. I think const static local variables would be fine. Save on additional string creation every time we hit this function.

File: src/ast/irbuilderbpf.cpp (hunk 1 above)
Comment (viktormalik) [reply src/ast/irbuilderbpf.cpp:2160]: It's not easy, I think. First, I wouldn't put it to semantic analyser but rather to a separate pass. And there, you'd have to introduce a context-sensitive analysis - you'd have to check that all field accesses of the pointer returned by the helper are done within a context where the pointer is not equal to NULL. I can't see a simple way of doing that.

Alternatively, we could inject the check ourselves into the AST or in codegen. But then, what would be the value for the other branch? Zero? That's a valid value for some fields so we would basically give users incorrect values.

The good thing here is that bpftrace will immediately inform you about the missing check so users don't need to remember anything, just add the check when the tool requires it.

File: man/adoc/bpftrace.adoc (hunk 4 above)
Comment (jordalgo) [reply man/adoc/bpftrace.adoc:1880]: What if NULL was an alias for `0`?

Comment (viktormalik) [reply man/adoc/bpftrace.adoc:1880]: That would be probably reasonable. I would do it in a separate PR, though.

File: tests/codegen/llvm/call_len_map_sum_elem_count.ll (hunk 6 above)
Comment (viktormalik) [reply tests/codegen/llvm/call_len_map_sum_elem_count.ll:120]: Interestingly yes, it looks like the verifier doesn't care about the name for function parameters as much as for global variables. In any case, now the name matches the name in kernel BTF.

File: man/adoc/bpftrace.adoc (hunk 3 above)
Comment (viktormalik) [reply man/adoc/bpftrace.adoc:1867]: Ouch, good catch.

File: src/ast/dibuilderbpf.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (viktormalik) [reply src/ast/dibuilderbpf.cpp:217]: Ok, will fix.

File: src/ast/irbuilderbpf.cpp (hunk 1 above)
Comment (danobi) [reply src/ast/irbuilderbpf.cpp:2160]: This seems like something we can improve on in the verifier. If the CPU number is statically known and in the range of valid CPUs, verifier should be able to let user omit null check.

File: man/adoc/bpftrace.adoc [hunk 8]
```diff
@@ -1846,6 +1850,39 @@ the path will be clamped by `size` otherwise `BPFTRACE_MAX_STRLEN` is used.
...
+
+The second variant may return NULL if `cpu` is higher than the number of
+available CPUs. Therefore, it is necessary to perform a NULL-check on the result
+when accessing fields of the pointed structure, otherwise the BPF program will
+be rejected.
+
+----
+interval:s:1 {
+  $runqueues = (struct rq *)percpu_kaddr("runqueues");
+  if ($runqueues != 0) {         # The check is mandatory here
```
Comment (danobi) [inline man/adoc/bpftrace.adoc:1880]: Comment in bpftrace is `//` or `/* */`, not `#`, right?

File: src/ast/irbuilderbpf.cpp [hunk 9]
```diff
@@ -2141,6 +2141,39 @@ CallInst *IRBuilderBPF::CreateGetFuncIp(Value *ctx, const location &loc)
...
+CallInst *IRBuilderBPF::CreateThisCpuPtr(Value *var, const location &loc)
+{
+  // void *bpf_per_cpu_ptr(const void *percpu_ptr)
+  // Return:
+  //    A pointer pointing to the kernel percpu variable on
+  //    this cpu.
```
Comment (danobi) [inline src/ast/irbuilderbpf.cpp:2166]: Maybe worth calling out "can never be NULL" here

File: src/ast/passes/codegen_llvm.cpp [hunk 10]
```diff
@@ -927,6 +927,18 @@ void CodegenLLVM::visit(Call &call)
...
+    auto var = b_.CreatePointerCast(DeclareKernelVar(name), b_.GET_PTR_TY());
+    Value *percpu_ptr;
+    if (call.vargs.size() == 1) {
+      percpu_ptr = b_.CreateThisCpuPtr(var, call.loc);
+    } else {
+      auto scoped_del = accept(call.vargs.at(1));
+      Value *cpu = expr_;
+      percpu_ptr = b_.CreatePerCpuPtr(var, cpu, call.loc);
+    }
+    expr_ = b_.CreatePtrToInt(percpu_ptr, b_.getInt64Ty());
```
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:941]: Is pointer to u64 right here? semantic analyser seems to mark the return type as u64

File: tests/semantic_analyser.cpp [hunk 11]
```diff
@@ -1366,6 +1366,24 @@ TEST(semantic_analyser, call_kaddr)
   test("kprobe:f { kaddr(123); }", 1);
 }
 
+TEST(semantic_analyser, call_percpu_kaddr)
+{
+  test("kprobe:f { percpu_kaddr(\"process_counts\"); }");
```
Comment (danobi) [inline tests/semantic_analyser.cpp:1371]: is `process_counts` from host? I don't see it in btf data source

File: man/adoc/bpftrace.adoc (hunk 8 above)
Comment (viktormalik) [reply man/adoc/bpftrace.adoc:1880]: Yes, of course :man_facepalming:

File: tests/semantic_analyser.cpp (hunk 11 above)
Comment (viktormalik) [reply tests/semantic_analyser.cpp:1371]: Yes, it's from the host. It exists on every Linux system but it's still probably better to have it in data_source.

File: src/ast/passes/codegen_llvm.cpp (hunk 10 above)
Comment (viktormalik) [reply src/ast/passes/codegen_llvm.cpp:941]: It creates a cast from a ptr to an integer, not to a "pointer to integer".

**Comment by viktormalik** [comment]
All comments should be addressed. Adding `process_counts` to `data_source` was quite a fight as pahole puts only percpu variables into BTF. So, I had to put the variable to the `.data..percpu` section and do some tweaks to building BTF and DWARF data in CMake (BTF needs relocatable file to preserve the custom section while DWARF needs executable file to be able to strip DWARF data into a separate file).

**Review by jordalgo — APPROVED** [review]
LGTM. Nice work!

**Review by danobi — APPROVED** [review]
yay, great to finally have this!

**Pull request merged**